### ISSUE TITLE
Correctly sfinae template constructors of IValue

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#34647]
Currently, IValue (`aten/src/ATen/core/ivalue.h`) has the following templated constructors:

```
    template<class T>
    IValue(c10::List<T> v);
    template<class T>
    IValue(at::ArrayRef<T> v);
    template<class T>
    IValue(const std::vector<T>& v);
```

These templated constructors recursively make use of the IValue(T) constructor, for whatever selection of what T is. Unfortunately, these constructors are not SFINAE'd, which means we will consider IValue constructible for *any* setting of T, even if the underlying IValue(T) constructor doesn't exist. This has lead to the ugly code here:

```
  // Assume T is decayed
  template <typename T>
  using not_ok_to_box =
    guts::disjunction<
      guts::negation<
        guts::disjunction<
          std::is_constructible<IValue, T>,
          // TensorOptions are not directly constructible into IValue,
          // but torch::jit::push knows how to handle them
          std::is_same<TensorOptions, T>,
          // void returns are ok
          std::is_same<void, T>
        >>
      ,   
      // some constructors are templated (and therefore pass
      // is_constructible), but do not actually work with all 
      // template arguments, so we must blacklist them explicitly
      // TODO: The correct fix is to sfinae based on is_constructible of T
      std::is_same<optional<ArrayRef<at::Dimname>>, T>,
      std::is_same<ArrayRef<at::Dimname>, T>
  
    >;
```

The purpose of this task is to appropriately SFINAE these constructors (and any other relevant constructors) with a test that IValue is constructible from T, so that the `not_ok_to_box` template can be simplified by removing the explicit tests for Dimname. You'll know you've done it right when everything compiles.

**Comment by ezyang** [comment]
This task is currently reserved for bootcamp, let me know if you have questions about it.

**Comment by terhuhf** [comment]
Closed by #34843
